### PR TITLE
ARROW-7172: [C++][Dataset] Improve format of Expression::ToString

### PR DESCRIPTION
**Pull request by bkietz**

### CODE REVIEW

File: cpp/src/arrow/dataset/filter.cc [hunk 1]
```diff
@@ -559,43 +558,35 @@ struct ScalarExpressionToString {
 };
 
 std::string ScalarExpression::ToString() const {
+  auto type_repr = value_->type->ToString();
```
Comment (nealrichardson) [inline cpp/src/arrow/dataset/filter.cc]: IDK that including this type is worthwhile (or if you think it is, then I probably want a different, simple print method that doesn't include it)

**Review by nealrichardson — COMMENTED** [review]
Basic idea looks good to me, just one comment. You'll of course need to edit the R test too.

File: cpp/src/arrow/dataset/filter.cc (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/filter.cc]: Even with implicit casts, the precise type of a scalar in an expression is significant. Notably: without implicit casts `"b"_ > 4:int32` might be invalid where `"b"_ > 4:float64` is not. If you're looking at these strings trying to debug a casting error, you really need the type information for the scalars. IMHO, it's out of scope to provide a print function without this; it seems like you might be planning to use these for something other than debugging (like direct re-evaluation in R, maybe)?

Comment (nealrichardson) [reply cpp/src/arrow/dataset/filter.cc]: I wasn't planning on that, no. My intended use case was more "what is in this object anyway?" and not "why did my object not evaluate?".

Comment (fsaintjacques) [reply cpp/src/arrow/dataset/filter.cc]: I would want the type to debug.

Comment (nealrichardson) [reply cpp/src/arrow/dataset/filter.cc]: 🤷‍♂ I'll take it, it's a huge improvement.

**Review by nealrichardson — APPROVED** [review]
LGTM from the R side; no opinion on the implementation.

**Comment by bkietz** [comment]
I'll wait on https://github.com/apache/arrow/pull/5857 to merge

**Comment by nealrichardson** [comment]
@bkietz ready for rebase and merge here

File: r/tests/testthat/test-expression.R [hunk 2]
```diff
@@ -54,9 +54,7 @@ test_that("C++ expressions", {
   expect_is(!(f < 4), "NotExpression")
   expect_output(
     print(f > 4),
-    # We can do better than this right?
-    # TODO: re-add the detail after ARROW-7172
-    'ComparisonExpression', #\nGREATER(field(f), scalar<double>(4))',
+    'ComparisonExpression\n(f > 4:double)',
```
Comment (pitrou) [inline r/tests/testthat/test-expression.R]: Why the `\n` by the way? It tends to make output more verbose and (IMHO) less readable in a prompt such as IPython.

Comment (bkietz) [reply r/tests/testthat/test-expression.R]: The class name and newline is only present in the R repr

**Pull request closed without merging**